readthedocs/builds/models.py

```diff
@@ -698,3 +699,114 @@ def run_time(self):
         if self.start_time is not None and self.end_time is not None:
             diff = self.end_time - self.start_time
             return diff.seconds
+
+
+class VersionAutomationRule(PolymorphicModel, TimeStampedModel):
+
+    """Versions automation rules for projects."""
+
+    REGEX_RULE = 'regex-rule'
+    RULE_TYPES = (
+        (REGEX_RULE, _('Regular expression rule')),
+    )
+
+    ACTIVATE_VERSION_ACTION = 'activate-version'
+    ACTIONS = (
+        (ACTIVATE_VERSION_ACTION, _('Activate version on match')),
+    )
+
+    project = models.ForeignKey(
+        Project,
+        related_name='automation_rules',
+        on_delete=models.CASCADE,
+    )
+    priority = models.IntegerField(_('Rule priority'))
+    rule_arg = models.CharField(
+        _('Value used for the rule to match the version'),
+        max_length=255,
+    )
+    action = models.CharField(
+        _('Action'),
+        max_length=32,
+        choices=ACTIONS,
+    )
+    action_arg = models.CharField(
+        _('Value used for the action to perfom an operation'),
+        max_length=255,
+        null=True,
+        blank=True,
+    )
+    version_type = models.CharField(
+        _('Version type'),
+        max_length=32,
+        choices=VERSION_TYPES,
+    )
+
+    class Meta:
+        unique_together = (('project', 'priority'),)
+        ordering = ('priority', '-modified', '-created')
+
+    def run(self, version, *args, **kwargs):
+        """
+        Run an action if `version` matches the rule.
+
+        :type version: readthedocs.builds.models.Version
+        """
+        if version.type == self.version_type:
+            match_result = self.match(version, self.rule_arg)
+            if match_result is not None:
+                return self.apply_action(version, match_result, self.action_arg)
+        return False
+
+    def match(self, version, arg):
+        """
+        Returns an object different from None if the version matches the rule.
```

Comment:
I think it's better to return `True` if it matches and `False` if there is no matching.

What is the `match_result` useful for?

Reply:
Actually, in case that `match_result` would be useful for something I'd suggest to have a specific method for that and leave `match` returning a boolean.

Reply:
We need to pass the match result, in case of regex, that would be used to capture the groups in a later phase of the feature.

Reply:
How exactly that "later phase of the feature" will use it?

Reply:
I mean, I don't see how the match result will be useful from a user's perspective. Or, how the user will have access to it.

In case we need it internally, how are we going to us it?